core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -699,76 +741,135 @@ private void allocateUnassigned() {
+
+                Decision currentDecision = allocation.deciders().canAllocate(shard, node.getRoutingNode(), allocation);
+                addNodeExplanation(nodeExplanationMap, node.getNodeId(), currentDecision, currentWeight);
+                if (currentDecision.type() == Type.YES || currentDecision.type() == Type.THROTTLE) {
+                    final boolean updateMinNode;
+                    if (currentWeight == minWeight) {
+                        /*  we have an equal weight tie breaking:
+                         *  1. if one decision is YES prefer it
+                         *  2. prefer the node that holds the primary for this index with the next id in the ring ie.
+                         *  for the 3 shards 2 replica case we try to build up:
+                         *    1 2 0
+                         *    2 0 1
+                         *    0 1 2
+                         *  such that if we need to tie-break we try to prefer the node holding a shard with the minimal id greater
+                         *  than the id of the shard we need to assign. This works find when new indices are created since
+                         *  primaries are added first and we only add one shard set a time in this algorithm.
+                         */
+                        if (currentDecision.type() == decision.type()) {
+                            final int repId = shard.id();
+                            final int nodeHigh = node.highestPrimary(shard.index().getName());
+                            final int minNodeHigh = minNode.highestPrimary(shard.getIndexName());
+                            updateMinNode = ((((nodeHigh > repId && minNodeHigh > repId)
+                                                   || (nodeHigh < repId && minNodeHigh < repId))
+                                                  && (nodeHigh < minNodeHigh))
+                                                 || (nodeHigh > minNodeHigh && nodeHigh > repId && minNodeHigh < repId));
+                        } else {
+                            updateMinNode = currentDecision.type() == Type.YES;
                         }
+                    } else {
+                        updateMinNode = true;
+                    }
+                    if (updateMinNode) {
+                        minNode = node;
+                        minWeight = currentWeight;
+                        decision = currentDecision;
                     }
                 }
             }
             if (decision == null) {
                 // decision was not set and a node was not assigned, so treat it as a NO decision
                 decision = Decision.NO;
             }
-            return Tuple.tuple(decision, minNode);
+            String explanation = null;
+            if (explain) {
+                if (decision.type() == Type.YES) {
+                    assert minNode != null;
+                    explanation = "shard assigned to node [" + minNode.getNodeId() + "]";
```

Comment:
same comment as above, calculating the explanation can be done outside of the balancer code.